src/phys/dca_step/cluster_solver/shared_tools/accumulation/tp/tp_accumulator_kernels.cu

```diff
@@ -179,24 +196,63 @@ __global__ void updateG4Kernel(CudaComplex<Real>* __restrict__ G4,
                                const CudaComplex<Real>* __restrict__ G_up, const int ldgu,
                                const CudaComplex<Real>* __restrict__ G_down, const int ldgd,
                                const int nb, const int nk, const int nw, const int nw_exchange,
-                               const int nk_exchange, const int sign, const bool atomic) {
+                               const int nk_exchange, const int sign, const bool atomic,
+                               const int my_rank, const int mpi_size,
+                               const uint64_t total_G4_size, bool distributed_g4_enabled) {
   // TODO: reduce code duplication.
   // TODO: decrease, if possible, register pressure. E.g. a single thread computes all bands.
 
-  const int size = nk * nw * nb * nb;
+  const uint64_t size = static_cast<uint64_t>(nk) * static_cast<uint64_t>(nw)
+                        * static_cast<uint64_t>(nb) * static_cast<uint64_t>(nb);
   // id_i is a linearized index of b1, b2, k1, k2.
-  const int id_i = blockIdx.x * blockDim.x + threadIdx.x;
   // id_j is a linearized index of b3, b4, k2, k_ex.
-  const int id_j = blockIdx.y * blockDim.y + threadIdx.y;
   // id_z is a linearized index of k_ex, w_ex.
-  const int id_z = blockIdx.z * blockDim.z + threadIdx.z;
-  if (id_i >= size || id_j >= size || id_z >= nw_exchange * nk_exchange)
-    return;
+  uint64_t id_i, id_j, id_z;
+  // global 1D linearized g4 index
+  uint64_t g4_index;
+
+  if(distributed_g4_enabled)
+  {
+   // With distributed g4 enabled, each rank only computes 1/total_mpi_ranks of G4. Specifically, the
+   // originally flattened one dimensional G4 array will be evenly divided by total number of
+   // mpi ranks into different regions (range length of each region is equal up to 1). Since
+   // each rank only allocates its own portion of G4, so offsetting index is needed. Each rank
+   // only computes G4 elements within correct starting and ending index range, otherwise, returns.
+    const uint64_t local_g4_index = static_cast<uint64_t>(blockIdx.x) * static_cast<uint64_t>(blockDim.x)
+                                    + static_cast<uint64_t>(threadIdx.x);
+
+    // get global G4 index
+    uint64_t start, end;
+    g4_helper.getComputeRange(my_rank, mpi_size, total_G4_size, start, end);
```

Comment:
there is no need to recompute start and end at every kernel launch. They can be computed once and passed as kernel parameters, or stored inside the g4_helper.